### DISCUSSION TITLE
Binary operators

### DISCUSSION THREAD

**Discussion by Mulugruntz** [geldata/gel#2358] [Ideas]
Hi!

Unless I'm wrong, I think EdgeQL doesn't allow for binary calculus (bitwise operators).

I'm thinking of:
```
edgedb> SELECT 0xFF0000 & 0x2342FB;
{2293760}
edgedb> SELECT hex(0xFF0000 & 0x2342FB);
{0x230000}
edgedb> SELECT 0o170000 & 0o223333;
{8192}
edgedb> SELECT oct(0o170000 & 0o223333);
{0o20000}
edgedb> SELECT bin(0x002800 ^ 0x010101);
{0b10010100100000001}
edgedb> SELECT 0xff0000 = 0b111111110000000000000000;
{true}
```
etc...

Did I miss some feature somewhere?
I was thinking of creating a bunch of computable properties for file permissions, directly in EdgeDB. So, I'd need some binary masking.

Else, is it already in discusssion? Needed? On the roadmap?

**Comment by elprans** [comment]
I believe we've discussed these a while ago and the consensus was that bit operators are too niche to deserve dedicated operators in a database.  That said, a set of functions implementing these operations on integral types would be a welcome addition.  Feel free to open an issue or a PR adding these.  This should be a simple matter of wrapping PostgreSQL's operators, e.g:

```
CREATE FUNCTION
std::bit_or(a: std::anyint, b: std::anyint) -> std::anyint
{
    CREATE ANNOTATION std::description := 'Bitwise OR of a and b';
    SET volatility := 'IMMUTABLE';
    USING SQL $$ SELECT "a" | "b" $$;
}
```

**Reply by Mulugruntz** [reply]
Interesting, I didn't think I could use PgSql directly in EdgeQL. :-)

**Reply by elprans** [reply]
Only in the standard library.
